metadata/src/main/java/org/apache/kafka/timeline/SnapshotRegistry.java

```diff
@@ -0,0 +1,247 @@
+     */
+    public Iterator<Snapshot> reverseIterator() {
+        return new ReverseSnapshotIterator();
+    }
+
+    /**
+     * Returns a sorted list of snapshot epochs.
+     */
+    public List<Long> epochsList() {
+        List<Long> result = new ArrayList<>();
+        for (Iterator<Snapshot> iterator = iterator(); iterator.hasNext(); ) {
+            result.add(iterator.next().epoch());
+        }
+        return result;
+    }
+
+    /**
+     * Gets the snapshot for a specific epoch.
+     */
+    public Snapshot getSnapshot(long epoch) {
+        Snapshot snapshot = snapshots.get(epoch);
+        if (snapshot == null) {
+            throw new RuntimeException("No snapshot for epoch " + epoch + ". Snapshot " +
+                "epochs are: " + epochsList().stream().map(e -> e.toString()).
+                    collect(Collectors.joining(", ")));
+        }
+        return snapshot;
+    }
+
+    /**
+     * Creates a new snapshot at the given epoch.
+     *
+     * @param epoch             The epoch to create the snapshot at.  The current epoch
+     *                          will be advanced to one past this epoch.
+     */
+    public Snapshot createSnapshot(long epoch) {
+        Snapshot last = head.prev();
+        if (last.epoch() >= epoch) {
+            throw new RuntimeException("Can't create a new snapshot at epoch " + epoch +
+                " because there is already a snapshot with epoch " + last.epoch());
+        }
+        Snapshot snapshot = new Snapshot(epoch);
+        last.appendNext(snapshot);
+        snapshots.put(epoch, snapshot);
+        log.debug("Creating snapshot {}", epoch);
+        Snapshot n = head.next();
+        while (n != head) {
+            n = n.next();
+        }
+        return snapshot;
+    }
+
+    /**
+     * Reverts the state of all data structures to the state at the given epoch.
+     *
+     * @param targetEpoch       The epoch of the snapshot to revert to.
+     */
+    public void revertToSnapshot(long targetEpoch) {
+        Snapshot target = getSnapshot(targetEpoch);
+        Iterator<Snapshot> iterator = iterator(target);
+        iterator.next();
+        while (iterator.hasNext()) {
+            Snapshot snapshot = iterator.next();
+            log.debug("Deleting snapshot {} because we are reverting to {}",
+                snapshot.epoch(), targetEpoch);
+            iterator.remove();
+        }
+        target.handleRevert();
+    }
+
+    /**
+     * Deletes the snapshot with the given epoch.
+     *
+     * @param targetEpoch       The epoch of the snapshot to delete.
+     */
+    public void deleteSnapshot(long targetEpoch) {
+        deleteSnapshot(getSnapshot(targetEpoch));
+    }
+
+    /**
+     * Deletes the given snapshot.
+     *
+     * @param snapshot          The snapshot to delete.
+     */
+    public void deleteSnapshot(Snapshot snapshot) {
+        Snapshot prev = snapshot.prev();
+        if (prev != head) {
+            prev.mergeFrom(snapshot);
```

Comment:
Not sure that I understand why we need to merge the deleted snapshot to previous snapshot. It seems that the previous snapshot should already have everything needed from the deleted snapshot?

Reply:
This is an optimization I made (I should have mentioned it earlier maybe).  I wanted PUT to be independent of the number of snapshot in existence.  So, PUT now adds a delta only to the latest snapshot, not to all snapshots where it is relevant.  That means that sometimes when reading a snapshot, we have to look at other snapshots to see if there are relevant entries in there.  Since reading from a snapshot is less common than modifying the current state, this seemed like a good tradeoff.

Reply:
So the merging logic is needed when deleting a snapshot that isn't the oldest snapshot